Add options for rails_cache_store

dgoetz opened the pull request:
This allows to make use of redis as backend, but still defaults to file.

@timogoebel: Can you have a look as the feature was implemented by you and so you perhaps know best if the parameters make sense in this way?

dgoetz commented:
I think the failed tests are unrelated, because it is just Debian failing for "install foreman with prometheus" which runs fine on CentOS.

dgoetz commented:
@ekohl I have it included now (also in the tests). Tests are still running.

ehelms commented on manifests/init.pp:
```diff
@@ -307,6 +310,7 @@
   Boolean $jobs_service_enable = $::foreman::params::jobs_service_enable,
   Boolean $hsts_enabled = $::foreman::params::hsts_enabled,
   Array[Stdlib::HTTPUrl] $cors_domains = $::foreman::params::cors_domains,
+  Hash $rails_cache_store = $::foreman::params::rails_cache_store,
```
Are the options for cache store well defined? I know from user issue reports that hashes are seen as complex and painful to send on the commandline when configuring through something like `foreman-installer`. If possible, breaking this out would be nice.

dgoetz replied:
I fear it is too flexible. Minimum are the following lines in the config which could also be omitted:
```
:rails_cache_store:
  :type: file
```
For redis it would be something like this as minimum:
```
:rails_cache_store:
  :type: redis
  :urls:
     - redis://localhost:8479/0
  :options:
    :compress: true
    :namespace: foreman
```
But it could be multiple URLs and more options.

In Puppet code it would look like:
```
{'type' => 'file'}
```
or:
```
{'type' => 'redis', 'urls' => ['localhost:8479/0'], options => {'compress' => 'true', 'namespace' => 'foreman'}}
```

Not sure how it would look for the installer. I am also not sure if future will perhaps bring more cache store backends and they will have a different format. 

With this in mind I do not come to a good alternative, best other way I can think of:
- rails_cache_store_type as string
- rails_cache_store_redis_urls as array of strings (only used when rails_cache_store_type equals redis)
- rails_cache_store_redis_options as array of strings which will be split on a delimiter (only used when rails_cache_store_type equals redis)

ekohl commented on manifests/init.pp:
```diff
@@ -208,6 +208,9 @@
 # $cors_domains::                 List of domains that show be allowed for Cross-Origin Resource Sharing. This requires Foreman 1.22+
 #
+# $rails_cache_store::            Set rails cache store. This requires Foreman 1.22+. Defaults to {'type' => 'file'}.
+#                                 For redis use {'type' => 'redis', 'urls' => ['localhost:8479/0'], options => {'compress' => 'true', 'namespace' => 'foreman'}}.
```
In another issue we've had the problem that this is confusing for foreman-installer users. How about an example in README?

dgoetz replied:
I added a paragraph to the README. Was this the one you wanted or did you mean the one of foreman-installer?

ekohl replied:
I think it can be dropped here now. You can also drop the 1.22+ part since the module is 2.0.0 only now.

ekohl replied:
Could you have a look at this?

dgoetz replied:
Shortened to "Set rails cache store" only

laugmanuel commented:
@dgoetz can I support you in any way on getting this merged? I would love to experiment with this in our release environment.

dgoetz commented:
@laugmanuel: No, it needs re-review and comments from @ekohl and @ehelms, so I get input for further adjustments if required.

dgoetz commented:
Rebased to resolve merge conflicts, tests failing on debian for statsd seem to be unrelated.

ekohl commented on manifests/install.pp:
```diff
@@ -48,4 +48,10 @@
     }
   }
+
+  if $::foreman::rails_cache_store['type'] == 'redis' {
```
If type is not present in the hash, I think this will fail and the type doesn't enforce it currently.

ekohl commented on manifests/params.pp:
```diff
@@ -188,6 +188,14 @@
+  # Rails Cache Store
+  $rails_cache_store = { 'type' => 'file' }
+
+  # KeepAlive settings of Apache
```
This looks like a result of a rebase with conflicts and can be dropped.

ekohl commented on README.md:
```diff
@@ -34,6 +34,19 @@ on the host this module is applied to. Databases will be created with using the
+containing the type and options specfic to the backend.
+
+The default is the file backend, configured via `{'type' => 'file'}`. To
+setup for redis use a hash similar to `{'type' => 'redis', 'urls' => ['localhost:8479/0'], options => {'compress' => 'true', 'namespace' => 'foreman'}}`
```
Would it make sense to add an example like:
```puppet
class { 'puppet':
  rails_cache_store => { 'type' => 'file' },
}
```

And also the longer example.

dgoetz commented:
I rebased and addressed the comments.
* I removed the additional lines from the rebase.
* An example was added.
* I tested how a hash missing a key is handled and it works fine (according to the puppet docs it is simple handled as undef).

ehelms commented:
@dgoetz This will need another rebase

dgoetz commented:
Rebased and tests are green except from beaker tests for puppet 5 testing journald, statsd and prometheus, not sure if this is related.

ekohl commented on manifests/init.pp:
```diff
@@ -301,6 +303,7 @@
   Integer[0] $foreman_service_puma_threads_min = $foreman::params::foreman_service_puma_threads_min,
   Integer[0] $foreman_service_puma_threads_max = $foreman::params::foreman_service_puma_threads_max,
   Integer[0] $foreman_service_puma_workers = $foreman::params::foreman_service_puma_workers,
+  Hash $rails_cache_store = $foreman::params::rails_cache_store,
```
```suggestion
  Hash[String, Any] $rails_cache_store = $foreman::params::rails_cache_store,
```

dgoetz replied:
Adjusted and pushed, tests are running again

dgoetz commented:
Tests are green now, and I hope I have addressed everything.

ekohl commented:
Thanks!
